src/main/java/seedu/recipe/ui/AddRecipeForm.java

```diff
@@ -0,0 +1,286 @@
+     * @return The created dynamic TextField.
+     */
+    private TextField createDynamicTextField(String text) {
+        TextField textField = new TextField(text);
+
+        //Keyboard listener for navigation
+        textField.addEventFilter(KeyEvent.KEY_PRESSED, event -> {
+            int currentIndex = ((VBox) textField.getParent()).getChildren().indexOf(textField);
+            if (event.getCode() == KeyCode.UP) {
+                // Condition 1: UP key pressed
+                if (currentIndex > 0) {
+                    TextField prevField = (TextField) ((VBox) textField.getParent())
+                                                                       .getChildren().get(currentIndex - 1);
+                    prevField.requestFocus();
+                }
+            } else if (event.getCode() == KeyCode.DOWN || event.getCode() == KeyCode.TAB) {
+                TextField nextField = (TextField) ((VBox) textField.getParent()).getChildren().get(currentIndex + 1);
+                // Condition 2.1: DOWN key pressed
+                if (event.getCode() == KeyCode.DOWN) {
+                    nextField.requestFocus();
+                // Condition 2.2: TAB key pressed
+                } else if (event.getCode() == KeyCode.TAB) {
+                    // If it is a new placeholder row and there's another TextField after it, skip to the field after
+                    if (nextField.getText().isEmpty() && currentIndex + 2
+                            < ((VBox) textField.getParent()).getChildren().size()) {
+                        nextField = (TextField) ((VBox) textField.getParent()).getChildren().get(currentIndex + 2);
+                    }
+                    nextField.requestFocus();
+                }
+                event.consume();
+            } else {
+                // Default: Do nothing
+            }
+        });
+        //Textfield listener for automatically adding/removing new input rows
+        textField.focusedProperty().addListener((observable, oldValue, newValue) -> {
+            VBox parentBox = (VBox) textField.getParent();
+            int lastIndex = parentBox.getChildren().size() - 1;
+            // Check if the TextField has gained focus
+            if (newValue) {
+                // Check if it's the last TextField in the VBox
+                if (parentBox.getChildren().indexOf(textField) == lastIndex) {
+                    TextField newField = createDynamicTextField("");
+                    parentBox.getChildren().add(newField);
+                }
+            } else {
+                // Check if it's the last TextField, it's empty, and the focus is not in the same VBox, then remove it
+                if (getNextTextField(textField).getText().isEmpty() && textField.getText().isEmpty()) {
+                    parentBox.getChildren().remove(getNextTextField(textField));
+                }
+            }
+        });
+        return textField;
+    }
```

Comment:
SLAP: you may want to abstract out each of the different key handlers into separate functions, e.g.
```java
    private TextField createDynamicTextField(String text) {
        TextField textField = new TextField(text);

        //Keyboard listener for navigation
        textField.addEventFilter(KeyEvent.KEY_PRESSED, event -> {
            int currentIndex = ((VBox) textField.getParent()).getChildren().indexOf(textField);
            if (event.getCode() == KeyCode.UP) {
                handleKeyCodeUp();
            } else if (event.getCode() == KeyCode.DOWN || event.getCode() == KeyCode.TAB) {
                handleKeyCodeDown();
            } else {
                // Default: Do nothing
            }
        });
        //Textfield listener for automatically adding/removing new input rows
        textField.focusedProperty().addListener((observable, oldValue, newValue) -> {
            handleRemoveNewInputRows();
        });
        return textField;
    }
```

Reply:
Thanks for providing the code example g!